include/dali/core/mm/monotonic_resource.h

```diff
@@ -0,0 +1,203 @@
+ public:
+  explicit monotonic_host_resource(memory_resource *upstream,
+                                   size_t next_block_size = 1024)
+  : upstream_(upstream), next_block_size_(next_block_size) {}
+
+  ~monotonic_host_resource() {
+    free_all();
+  }
+
+  void free_all() {
+    while (curr_block_) {
+      assert(curr_block_->sentinel == sentinel_value && "Memory corruption detected");
+      auto *prev = curr_block_->prev;
+      auto *base = reinterpret_cast<char*>(curr_block_) - curr_block_->usable_size;
+      size_t alloc_size = curr_block_->usable_size + sizeof(block_info);
+      upstream_->deallocate(base, alloc_size, curr_block_->alignment);
+      curr_block_ = prev;
+    }
+  }
+
+ private:
+  void *do_allocate(size_t bytes, size_t alignment) override {
+    char *ret = detail::align_ptr(curr_, alignment);
+    if (ret + bytes > limit_) {
+      while (next_block_size_ < bytes + sizeof(block_info))
+        next_block_size_ += next_block_size_;
+      alignment = std::max(alignment, alignof(block_info));
+      curr_ = static_cast<char*>(upstream_->allocate(next_block_size_, alignment));
+      limit_ = curr_ + next_block_size_ - sizeof(block_info);
+      block_info *prev = curr_block_;
+      if (prev) {
+        assert(prev->sentinel == sentinel_value && "Memory corruption detected");
+      }
+      curr_block_ = reinterpret_cast<block_info*>(limit_);
+      curr_block_->sentinel = sentinel_value;
+      curr_block_->prev = prev;
+      curr_block_->usable_size = next_block_size_ - sizeof(block_info);
+      curr_block_->alignment = alignment;
+      next_block_size_ += next_block_size_;
+
+      ret = curr_;
+    }
+
+    curr_ = ret + bytes;
+    return ret;
+  }
+
+  // don't deallocate at all
+  void do_deallocate(void *data, size_t bytes, size_t alignment) override {
+  }
+
+  static constexpr size_t sentinel_value = detail::sentinel_value<size_t>::value;
+
+  char *curr_ = nullptr, *limit_ = nullptr;
+
+  memory_resource *upstream_;
+  size_t next_block_size_;
+  struct block_info {
+    size_t sentinel;
+    block_info *prev;
+    size_t usable_size;
+    size_t alignment;
+  };
+  block_info *curr_block_ = nullptr;
+};
+
+/**
+ * @brief Monotonic resource that keeps separate host memory to manage upstream blocks.
+ *
+ * Use this resource when the upstream memory cannot be accessed by the main processor for the
+ * purpose of storing the memory management structures.
+ *
+ * Monotonic resources don't require manual deallocation of memory.
+ * The lifetime of a monotnic resource is limited and all memory will be deallocated in bulk
+ * when the resource is destroyed.
+ */
+class monotonic_device_resource : public memory_resource {
+ public:
+  explicit monotonic_device_resource(memory_resource *upstream,
+                                    size_t next_block_size = 1024)
+  : upstream_(upstream), next_block_size_(next_block_size) {}
+
+  ~monotonic_device_resource() {
```

Comment:
```suggestion
  ~monotonic_device_resource() override {
```

Reply:
We don't care.

Reply:
For me, it's the way to tell "I am expecting this destructor to be called when deleting a pointer of the base class". If you miss the virtual on the base class you'll get an error. Your call